### PR TITLE
Reworks multitool buffer into a component

### PR DESCRIPTION
**Pull request by PowerfulBacon**
## About The Pull Request

**This PR is a dependency for my upcoming mining refactor.**

Reworks the multitool buffer into a component.
This component can be applied to any object.
You can listen for the COMSIG_PARENT_RECIEVE_BUFFER on a target to see when someone is using a buffer object on your object.
You can send COMSIG_ITEM_PUSH_BUFFER to push an item to the buffer.
You can send COMSIG_ITEM_FLUSH_BUFFER to flush the buffer.

Adds some defines for quickly adding in the buffer handlers, cause I am lazy but they can be expanded if requested since they might be considered a bit ugly.

Dependancy of my upcoming lavaland rework.

## Why It's Good For The Game

Removes some hard-deletes from multitool buffer.

## Testing Photographs and Procedure

Provided below is successful testing documentation for every single object that uses the multitool's buffer feature.

<details>
<summary>
Ore silo + Remote Material Containers
</summary>

image
image
image

</details>

<details>
<summary>
Cloning
</summary>

image
Didn't even know this was a think to be honest.

</details>

<details>
<summary>
Operating Computer + Stasis Beds
</summary>

image
image

I have since fixed the issue with hitting the computer during linking

</details>

<details>
<summary>
Launchpad
</summary>

image

</details>

<details>
<summary>
Portable Turret Controller + Turret Cover
</summary>

image

</details>

<details>
<summary>
Telecommunications
</summary>

image
image
image
image
image

</details>

<details>
<summary>
Teleporter
</summary>

image
Hitting has since been fixed.
image
image

</details>

<details>
<summary>
Pirate Console
</summary>

image
image

</details>

<details>
<summary>
Xenobiology (Monkey Recycler + Camera)
</summary>

image

</details>

<details>
<summary>
Processing Unit
</summary>

image
Runtime has since been fixed.

</details>

<details>
<summary>
Stacking Machine
</summary>

image
Made this one slightly more intuative by letting it have links on both ways.

</details>

<details>
<summary>
Conveyor Belt + Switch
</summary>

image

</details>

<details>
<summary>
Bluespace Artillery
</summary>

image

Why does this even have multitool linking?

</details>

## Changelog
:cl:
code: Refactors multitool buffers to use a component, removing some hard-delete potential.
/:cl:

### CODE REVIEW

File: code/game/machinery/cloning.dm [hunk 1]
```diff
@@ -404,6 +404,25 @@
...
+			return
+		to_chat(user, "<font color = #666633>-% Successfully linked [buffer] with [src] %-</font color>")
+		var/obj/machinery/computer/cloning/comp = buffer
+		if(connected)
+			connected.DetachCloner(src)
+		comp.AttachCloner(src)
+	else if (TRY_STORE_IN_BUFFER(buffer_parent, src))
+		to_chat(user, "<font color = #666633>-% Successfully stored [REF(src)] [name] in buffer %-</font color>")
+	else
+		return NONE
```
Comment (itsmeow) [inline code/game/machinery/cloning.dm:423]: the returning of COMPONENT_NO_AFTERATTACK, null, FALSE, and NONE is relatively inconsistent in these procs

could it be standardized?

Comment (PowerfulBacon) [reply code/game/machinery/cloning.dm:423]: I'll make any signals that want to return no result 0 and any signal which does not expect return values not have return (null)

File: code/datums/components/buffer.dm [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+	return COMPONENT_NO_INTERACT
+
+/datum/component/buffer/proc/populate_buffer(datum/source, datum/buffer_entity)
+	SIGNAL_HANDLER
+	if (target)
+		UnregisterSignal(target, COMSIG_PARENT_QDELETING)
+	target = buffer_entity
+	if (target)
+		RegisterSignal(target, COMSIG_PARENT_QDELETING, PROC_REF(flush_buffer))
+		return COMPONENT_BUFFER_STORE_SUCCESS
```
Comment (PowerfulBacon) [inline code/datums/components/buffer.dm:54]:
```suggestion
		return COMPONENT_BUFFER_STORE_SUCCESS
	return NONE
```

File: code/game/machinery/cloning.dm (hunk 1 above)
Comment (PowerfulBacon) [reply code/game/machinery/cloning.dm:423]: Oh yea that was super inconsistent

File: code/game/machinery/computer/launchpad_control.dm [hunk 3]
```diff
@@ -113,22 +113,20 @@
...
-				ui_update()
-				to_chat(user, "<span class='notice'>You upload the data from the [W.name]'s buffer.</span>")
-			else
-				to_chat(user, "<span class='warning'>[src] cannot handle any more connections!</span>")
-	else
-		return ..()
+REGISTER_BUFFER_HANDLER(/obj/machinery/computer/launchpad)
+
+DEFINE_BUFFER_HANDLER(/obj/machinery/computer/launchpad)
+	if(istype(buffer, /obj/machinery/launchpad))
```
Comment (itsmeow) [inline code/game/machinery/computer/launchpad_control.dm:119]: early return

File: code/game/machinery/quantum_pad.dm [hunk 4]
```diff
@@ -103,6 +82,26 @@
 
 	return ..()
 
+REGISTER_BUFFER_HANDLER(/obj/machinery/quantumpad)
+
+DEFINE_BUFFER_HANDLER(/obj/machinery/quantumpad)
+	if (panel_open)
```
Comment (itsmeow) [inline code/game/machinery/quantum_pad.dm:88]: early return

**Review by itsmeow — CHANGES_REQUESTED** [review]
a looot of these can use early returns since it amounts to if(x) do a bunch else return none

**Comment by PowerfulBacon** [comment]
Most of these are just copy and pastes of the existing behaviours to not risk any accidental changes

File: code/game/machinery/quantum_pad.dm (hunk 4 above)
Comment (PowerfulBacon) [reply code/game/machinery/quantum_pad.dm:88]: This is an early return

**Review by itsmeow — APPROVED** [review]

**Pull request merged**